doc/design/programming_solver.md

```diff
@@ -0,0 +1,141 @@
+
+- [Linear programming](https://en.wikipedia.org/wiki/Linear_programming)
+- [Quadratic programming](https://en.wikipedia.org/wiki/Quadratic_programming)
+- [Nonlinear programming](https://en.wikipedia.org/wiki/Nonlinear_programming)
+- [Integer programming](https://en.wikipedia.org/wiki/Integer_programming)
+- [Mixed-integer linear programming](http://macc.mcmaster.ca/maccfiles/chachuatnotes/07-MILP-I_handout.pdf)
+- [Mixed-integer quadratic programming](http://www.optimization-online.org/DB_FILE/2014/07/4446.pdf)
+- [Mixed-integer nonlinear programming](https://link.springer.com/book/10.1007/978-1-4614-1927-3)
+
+## Background
+
+To understand what's a mathematical programming problem, let's take a look at this example:
+
+Giapetto’s Woodcarving, Inc., manufactures two types of wooden toys: soldiers and trains. 
+
+A soldier sells for $27 and uses $10 worth of raw materials.  Each soldier that is manufactured increases Giapetto’s variable labor and overhead costs by $14.  A train sells for $21 and uses $9 worth of raw materials.  Each train built increases Giapetto’s variable labor and overhead costs by $10.  The manufacture of wooden soldiers and trains requires two types of skilled labor: carpentry and finishing.  A soldier requires 2 hours of finishing labor and 1 hour of carpentry labor.  A train requires 1 hour of finishing labor and 1 hour of carpentry labor.  Each week, Giapetto can obtain all the needed raw material but only 100 finishing hours and 80 carpentry hours.  Demand for trains is unlimited, but at most 40 soldiers are bought each week.  Giapetto wants to maximize weekly profit (revenues-costs).
+
+Let
+
+- x be the number of soldiers produced each week
+- y be the number of trains produced each week
+
+Then the objective is: 
+
+**Maximise Z = (27 - 10 - 14)x + (21 - 9 - 10)y = 3x + 2y**
+
+The constraints are:
+
+- 2x + y <= 100 (Finishing Constraint)
+- x + y <= 80 (Carpentry Constraint)
+- x <=40 (Constraint on demand for soldiers)
+- x,y >= 0 (sign restriction)
+
+
+To solve this problem, we can write objective and constraints using SQL, then SQLFlow should call the corresponding "solver" (e.g. [pyomo](http://www.pyomo.org/) and [GLPK](https://www.gnu.org/software/glpk/)) to find the result and output the result into the result table.
+
+## Grammar Design
+
+Generally, SQLFlow users can write one SQL statement to describe the programming problem:
+
+```sql
+SELECT * FROM train_table
+TO SOLVE LP
+WITH
+     objective="sum([(train_table.price[i] - train_table.materials_cost[i] - train_table.other_cost[i]) * @X[i] for i in @X])",
+     constraints=["sum([train_table.finishing[i] * @X[i] for i in @X]) <= 100",
+                  "sum([train_table.carpentry[i] * @X[i] for i in @X]) <= 80",
+                  "@X[i] <= train_table.max_num[i] for i in @X",
```

Comment:
> @X[i] <= train_table.max_num[i] for i in @X

It's not a complete Python code, do you mean

``` python
all(@X[i] <= train_table.max_num[i] for i in @X") == True
```

Reply:
This is the syntax that `pyomo` accepts.

Reply:
Actually  syntax like '@x[i] <= train_table.max_num[i] for i in @x' is  not the pyomo available syntax, but it simplifies the syntax and converts to pyomo syntax backend.